crates/egui/src/ui_stack.rs

```diff
@@ -0,0 +1,157 @@
+                | Self::RightPanel
+                | Self::TopPanel
+                | Self::BottomPanel
+        )
+    }
+}
+
+// ----------------------------------------------------------------------------
+
+/// Information about a [`crate::Ui`] to be included in the corresponding [`UiStack`].
+#[derive(Default, Copy, Clone, Debug)]
+pub struct UiStackInfo {
+    pub kind: Option<UiKind>,
+    pub frame: Frame,
+}
+
+// ----------------------------------------------------------------------------
+
+/// Information about a [`crate::Ui`] and its parents.
+///
+/// [`UiStack`] serves to keep track of the current hierarchy of [`crate::Ui`]s, such
+/// that nested widgets or user code may adapt to the surrounding context or obtain layout information
+/// from a [`crate::Ui`] that might be several steps higher in the hierarchy.
+///
+/// Note: since [`UiStack`] contains a reference to its parent, it is both a stack, and a node within
+/// that stack. Most of its methods are about the specific node, but some methods walk up the
+/// hierarchy to provide information about the entire stack.
+#[derive(Clone, Debug)]
+pub struct UiStack {
+    // stuff that `Ui::child_ui` can deal with directly
+    pub id: Id,
+    pub kind: Option<UiKind>,
+    pub frame: Frame,
+    pub layout_direction: Direction,
+    pub min_rect: Rect,
+    pub max_rect: Rect,
+    pub parent: Option<Arc<UiStack>>,
+}
+
+// these methods act on this specific node
+impl UiStack {
+    /// Is this [`crate::Ui`] a panel?
+    #[inline]
+    pub fn is_panel_ui(&self) -> bool {
+        self.kind.map_or(false, |kind| kind.is_panel())
+    }
+
+    /// Is this a root [`crate::Ui`], i.e. created with [`crate::Ui::new()`]?
+    #[inline]
+    pub fn is_root_ui(&self) -> bool {
+        self.parent.is_none()
+    }
+
+    /// This this [`crate::Ui`] a [`crate::Frame`] with a visible stroke?
+    #[inline]
+    pub fn has_visible_frame(&self) -> bool {
+        self.frame.stroke.width > 0.0 && self.frame.stroke.color != Color32::TRANSPARENT
+    }
+}
+
+// these methods act on the entire stack
+impl UiStack {
+    /// Return an iterator that walks the stack from this node to the root.
+    #[allow(clippy::iter_without_into_iter)]
+    pub fn iter(&self) -> UiStackIterator {
+        UiStackIterator {
+            next: Some(Arc::new(self.clone())),
+        }
+    }
+
+    /// Check if this node is or is contained in a [`crate::Ui`] of a specific kind.
+    pub fn contained_id(&self, kind: UiKind) -> bool {
+        self.iter().any(|frame| frame.kind == Some(kind))
+    }
+}
+
+// ----------------------------------------------------------------------------
+
+/// Iterator that walks up a stack of `StackFrame`s.
+///
+/// See [`UiStack::iter`].
+pub struct UiStackIterator {
+    next: Option<Arc<UiStack>>,
+}
+
+impl Iterator for UiStackIterator {
+    type Item = Arc<UiStack>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let current = self.next.clone();
```

Comment:
If you use references (and lifetimes) we can avoid this `.clone()`. I think this is likely to be on the hot path of quite a few layout thing in the future.